metaflow/plugins/aws/batch/batch_decorator.py

```diff
@@ -292,9 +312,68 @@ def task_finished(
             # Best effort kill
             pass
 
+        if is_task_ok and getattr(flow, "_control_mapper_tasks", []):
+            self._wait_for_mapper_tasks(flow, step_name)
+
+    def _wait_for_mapper_tasks(self, flow, step_name):
+        """
+        When lauching multinode task with UBF, need to wait for the secondary
+        tasks to finish cleanly and produce their output before exiting the
+        main task. Otherwise main task finishing will cause secondary nodes
+        to terminate immediately, and possibly prematurely.
+        """
+        from metaflow import Step  # avoid circular dependency
+
+        t = time.time()
+        TIMEOUT = 600
+        print("Waiting for batch secondary tasks to finish")
+        while t + TIMEOUT > time.time():
+            time.sleep(2)
+            try:
+                step_path = "%s/%s/%s" % (flow.name, current.run_id, step_name)
+                tasks = [task for task in Step(step_path)]
+                if len(tasks) == len(flow._control_mapper_tasks) - 1:
+                    if all(
+                        task.finished_at is not None for task in tasks
+                    ):  # for some reason task.finished fails
+                        return True
+                else:
+                    print(
+                        "Not sufficient number of tasks:",
+                        len(tasks),
+                        len(flow._control_mapper_tasks),
+                    )
+            except Exception as e:
+                print(e)
+                pass
+        raise Exception(
+            "Batch secondary workers did not finish in %s seconds" % TIMEOUT
+        )
+
     @classmethod
     def _save_package_once(cls, flow_datastore, package):
         if cls.package_url is None:
             cls.package_url, cls.package_sha = flow_datastore.save_data(
                 [package.blob], len_hint=1
             )[0]
+
+
+def _setup_multinode_environment():
```

Comment:
we could add properties of `multinode_environment` to `current` too. Like here https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/project_decorator.py#L54

we can add them under a key `parallel`, so one can do e.g. `current.parallel.main_ip`:
```
Parallel = namedtuple('Parallel', ['main_ip', 'num_nodes'])
current._update_env({'parallel': Parallel(main_ip='..', num_nodes=2)})
```

Reply:
also we could add these to `metadata` like here https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/aws/step_functions/step_functions_decorator.py#L28

Reply:
added to current.